features/cellular/framework/targets/UBLOX/AT/UBLOX_AT_CellularStack.cpp

```diff
@@ -0,0 +1,582 @@
+    _at.unlock();
+
+    if (_at.get_last_error() == NSAPI_ERROR_OK) {
+        socket->remoteAddress = addr;
+        socket->connected = true;
+        return NSAPI_ERROR_OK;
+    }
+
+    return NSAPI_ERROR_NO_CONNECTION;
+}
+
+nsapi_size_or_error_t UBLOX_AT_CellularStack::socket_sendto_impl(CellularSocket *socket, const SocketAddress &address,
+        const void *data, nsapi_size_t size)
+{
+    int sent_len = 0;
+    uint8_t ch = 0;
+
+    _at.lock();
+    if (socket->proto == NSAPI_UDP) {
+        if (size > (nsapi_size_t) get_max_packet_size()) {
+            return NSAPI_ERROR_PARAMETER;
+        }
+        _at.cmd_start("AT+USOST=");
+        _at.write_int(socket->id);
+        _at.write_string(address.get_ip_address(), true);
+        _at.write_int(address.get_port());
+        _at.write_int(size);
+        _at.cmd_stop();
+        wait_ms(50);
+        while (ch != '@') {
+            _at.read_bytes(&ch, 1);
+        }
+        _at.write_bytes((uint8_t *)data, size);
+
+        _at.resp_start("+USOST:");
+        _at.skip_param(); // skip socket id
+        sent_len = _at.read_int();
+        _at.resp_stop();
+
+        if ((_at.get_last_error() == NSAPI_ERROR_OK)) {
+            return sent_len;
+        } else {
+            socket->rx_avail = false;
+        }
+    } else if (socket->proto == NSAPI_TCP) {
+        bool success = true;
+        const char *buf = (const char *) data;
+        nsapi_size_t blk = UBLOX_MAX_PACKET_SIZE;
+        nsapi_size_t count = size;
+
+        while ((count > 0) && success) {
+            if (count < blk) {
+                blk = count;
+            }
+            _at.cmd_start("AT+USOWR=");
+            _at.write_int(socket->id);
+            _at.write_int(blk);
+            _at.cmd_stop();
+            wait_ms(50);
+            while (ch != '@') {
+                _at.read_bytes(&ch, 1);
+            }
+            _at.write_bytes((uint8_t *)buf, blk);
+
+            _at.resp_start("+USOWR:");
+            _at.skip_param(); // skip socket id
+            sent_len = _at.read_int();
+            _at.resp_stop();
+
+            if ((sent_len >= (int) blk) &&
+                (_at.get_last_error() == NSAPI_ERROR_OK)) {
+            } else {
+                success = false;
+            }
+
+            buf += blk;
+            count -= blk;
+        }
+
+        if (success && _at.get_last_error() == NSAPI_ERROR_OK) {
+            socket->rx_avail = false;
+            return size - count;
+        }
+    }
+    _at.unlock();
+
+    return _at.get_last_error();
+}
+
+nsapi_size_or_error_t UBLOX_AT_CellularStack::socket_recvfrom_impl(CellularSocket *socket, SocketAddress *address,
+        void *buffer, nsapi_size_t size)
+{
+    nsapi_size_or_error_t nsapi_error_size = NSAPI_ERROR_DEVICE_ERROR;
+    bool success = true;
+    nsapi_size_t read_blk;
+    nsapi_size_t count = 0;
+    nsapi_size_t usorf_sz;
+    char ipAddress[NSAPI_IP_SIZE];
+    uint8_t ch = 0;
+    int port = 0;
+    Timer timer;
+
```

Comment:
To support async sockets, here should be something like this:
`    if (!socket->rx_avail) {
        _at.process_oob();
        if (!socket->rx_avail) {
            return NSAPI_ERROR_WOULD_BLOCK;
        }
    }
`

Reply:
We are using 'pending_bytes' for received data and already checking '_at.process_oob()' and returning 'NSAPI_ERROR_WOULD_BLOCK' if nothing is received.

Reply:
An application may call `recvfrom` even when there is no incoming data. In these cases function should return immediately with NSAPI_ERROR_WOULD_BLOCK (async) instead of waiting for `timer.read_ms() < SOCKET_TIMEOUT)`.

Reply:
What if data is received after some delay, as I have experienced in TCP case, data is received after some seconds of delay thats why I have added Timer to check availability of data until timer expires.

Reply:
You can always return `NSAPI_ERROR_WOULD_BLOCK`:

In **blocking** mode Mbed OS socket layer waits for a message until socket-timeout and if one is received via URC then socket layer calls `recvfrom` again to actually read the message.

In **non-blocking** mode control returns immediately to a calling application. When message is received via URC it issues callback on an application, then the application needs to call `recvfrom` again to actually read the message.

Reply:
Updated and removed `detach()` function from ublox child class and added 50ms wait in `test_detach()` test case to process urc's, received after detach and also used `mbed_poll` to wait for input.